crates/engine/tree/src/tree/root.rs

```diff
@@ -464,31 +464,56 @@ where
     }
 }
 
+/// Returns accounts only with those storages that were not already fetched, and
+/// if there are no such storages and the account itself was already fetched, the
+/// account shouldn't be included.
 fn get_proof_targets(
     state_update: &HashedPostState,
-    fetched_proof_targets: &HashMap<B256, HashSet<B256>>,
+    fetched_proof_targets: &mut HashMap<B256, HashSet<B256>>,
 ) -> HashMap<B256, HashSet<B256>> {
-    state_update
-        .accounts
-        .keys()
-        .filter(|hashed_address| !fetched_proof_targets.contains_key(*hashed_address))
-        .map(|hashed_address| (*hashed_address, HashSet::default()))
-        .chain(state_update.storages.iter().map(|(hashed_address, storage)| {
-            let fetched_storage_proof_targets = fetched_proof_targets.get(hashed_address);
-            (
-                *hashed_address,
-                storage
-                    .storage
-                    .keys()
-                    .filter(|slot| {
-                        !fetched_storage_proof_targets
-                            .is_some_and(|targets| targets.contains(*slot))
-                    })
-                    .copied()
-                    .collect(),
-            )
-        }))
-        .collect()
+    let mut targets = HashMap::default();
+
+    // first collect all new accounts (not previously fetched)
+    for &hashed_address in state_update.accounts.keys() {
+        if let Entry::Vacant(_) = fetched_proof_targets.entry(hashed_address) {
+            targets.insert(hashed_address, HashSet::default());
+        }
+    }
+
+    // then process storage slots for all accounts in the state update
+    for (hashed_address, storage) in &state_update.storages {
+        // only process storage if account:
+        // 1. is new (in targets), or
+        // 2. was previously fetched (in fetched_proof_targets), or
+        // 3. has storage updates even if not modified
+        if targets.contains_key(hashed_address) ||
+            fetched_proof_targets.contains_key(hashed_address) ||
+            state_update.storages.contains_key(hashed_address)
```

Comment:
why is this check necessary at all? this will always be `true` because we are iterating over `state_update.storages`

Reply:
you are right, the condition is always true now, removed